### PR TITLE
⚡️ Speed up function `fix_unescaped_quotes_in_json` by 106%

### PR DESCRIPTION
**Pull request by misrasaurabh1**
### 📄 106% (1.06x) speedup for ***`fix_unescaped_quotes_in_json` in `skyvern/forge/sdk/api/llm/utils.py`***

⏱️ Runtime :   **`13.1 milliseconds`**  **→** **`6.37 milliseconds`** (best of `320` runs)
<details>
<summary> 📝 Explanation and details</summary>

To optimize the provided Python function for faster processing, we can minimize the number of list operations, especially the sorting and appending operations. Instead of appending indices and adding escape characters afterward, we can build the output string directly during the traversal of the input string, which will reduce the need for post-processing.

### Changes Made.
1. **Avoided List Sorting and Index Shifting**: Instead of creating a list of indices and performing additional operations to add escape characters, we directly construct the final string during the traversal.
2. **Efficient String Construction**: Use a list to accumulate characters to avoid the inefficiencies of repeatedly concatenating strings.
3. **Single Pass Approach**: Ensured all operations are done within a single pass through the input string to optimize runtime.

These changes will improve the overall performance and maintain the correctness of the function.

</details>

✅ **Correctness verification report:**

| Test                        | Status            |
| --------------------------- | ----------------- |
| ⚙️ Existing Unit Tests | 🔘 **None Found** |
| 🌀 Generated Regression Tests | ✅ **52 Passed** |
| ⏪ Replay Tests | 🔘 **None Found** |
| 🔎 Concolic Coverage Tests | 🔘 **None Found** |
|📊 Tests Coverage       | 100.0% |
<details>
<summary>🌀 Generated Regression Tests Details</summary>

```python
import pytest  # used for our unit tests
# function to test
import structlog
from skyvern.forge.sdk.api.llm.utils import fix_unescaped_quotes_in_json

LOG = structlog.get_logger()
from skyvern.forge.sdk.api.llm.utils import fix_unescaped_quotes_in_json

# unit tests

def test_basic_valid_json():
    # Basic valid JSON strings that should remain unchanged
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key1": "value1", "key2": "value2"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"nested": {"key": "value"}}')

def test_escaped_quotes():
    # JSON strings with escaped quotes that should remain unchanged
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with \\"escaped\\" quotes"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "another \\"example\\" with escapes"}')

def test_unescaped_quotes():
    # JSON strings with unescaped quotes that need fixing
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with "unescaped" quotes"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "another "example" with unescaped quotes"}')

def test_mixed_quotes():
    # JSON strings with mixed escaped and unescaped quotes
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with \\"escaped\\" and "unescaped" quotes"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "another "example" with \\"escaped\\" quotes"}')

def test_nested_structures():
    # JSON strings with nested structures
    codeflash_output = fix_unescaped_quotes_in_json('{"outer": {"inner": "value with "unescaped" quotes"}}')
    codeflash_output = fix_unescaped_quotes_in_json('{"array": ["value with "unescaped" quotes", "another value"]}')

def test_whitespace():
    # JSON strings with various amounts of whitespace
    codeflash_output = fix_unescaped_quotes_in_json('{"key" : "value with "unescaped" quotes" }')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with "unescaped" quotes" , "another_key" : "another value"}')

def test_edge_cases():
    # Edge cases with empty or minimal content
    codeflash_output = fix_unescaped_quotes_in_json('')
    codeflash_output = fix_unescaped_quotes_in_json('{}')
    codeflash_output = fix_unescaped_quotes_in_json('[]')
    codeflash_output = fix_unescaped_quotes_in_json(' ')
    codeflash_output = fix_unescaped_quotes_in_json('\n\t')

def test_invalid_json():
    # Invalid JSON strings that should still be processed
    codeflash_output = fix_unescaped_quotes_in_json('{"key": value with "unescaped" quotes}')
    codeflash_output = fix_unescaped_quotes_in_json('{key: "value with "unescaped" quotes"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with "unescaped" quotes", "another_key": value}')

def test_large_scale():
    # Large scale test cases
    large_json = '{"' + '", "'.join([f'key{i}": "value with "unescaped" quotes"' for i in range(1000)]) + '}'
    expected_large_json = '{"' + '", "'.join([f'key{i}": "value with \\"unescaped\\" quotes"' for i in range(1000)]) + '}'
    codeflash_output = fix_unescaped_quotes_in_json(large_json)

    nested_json = '{"outer": ' * 10 + '{"key": "value with "unescaped" quotes"}' + '}' * 10
    expected_nested_json = '{"outer": ' * 10 + '{"key": "value with \\"unescaped\\" quotes"}' + '}' * 10
    codeflash_output = fix_unescaped_quotes_in_json(nested_json)

def test_complex_structures():
    # Complex JSON structures
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value", "array": [{"nested_key": "nested value with "unescaped" quotes"}], "boolean": true}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value", "null_value": null, "number": 123, "nested": {"inner_key": "inner value with "unescaped" quotes"}}')

def test_special_characters():
    # JSON strings with special characters and escape sequences
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with newline\n and "unescaped" quotes"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with tab\t and "unescaped" quotes"}')

def test_unicode_characters():
    # JSON strings with Unicode characters
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with unicode \u1234 and "unescaped" quotes"}')
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with emoji 😊 and "unescaped" quotes"}')

if __name__ == "__main__":
    pytest.main()
# codeflash_output is used to check that the output of the original code is the same as that of the optimized code.

import pytest  # used for our unit tests
import structlog
from skyvern.forge.sdk.api.llm.utils import fix_unescaped_quotes_in_json

# function to test
LOG = structlog.get_logger()
from skyvern.forge.sdk.api.llm.utils import fix_unescaped_quotes_in_json

# unit tests

# Basic Valid JSON Strings
def test_simple_json_object():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value"}')

def test_nested_json_object():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": {"nested_key": "nested_value"}}')

def test_json_array():
    codeflash_output = fix_unescaped_quotes_in_json('["value1", "value2", "value3"]')

# JSON Strings with Unescaped Quotes Inside Strings
def test_single_unescaped_quote():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with unescaped "quote" inside"}')

def test_multiple_unescaped_quotes():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with "multiple" unescaped "quotes" inside"}')

def test_unescaped_quotes_different_positions():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "start "middle" end"}')

# JSON Strings with Escaped Quotes Inside Strings
def test_properly_escaped_quotes():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with escaped \\"quote\\" inside"}')

def test_mixed_escaped_unescaped_quotes():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with escaped \\"quote\\" and unescaped "quote" inside"}')

# JSON Strings with Special Characters
def test_special_characters_inside_strings():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with special characters !@#$%^&*() inside"}')

def test_unicode_characters():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with unicode characters \\u1234 inside"}')

# JSON Strings with Whitespace
def test_whitespace_around_quotes():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with unescaped " quote " inside"}')

def test_whitespace_inside_strings():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": "value with    spaces   inside"}')

# Edge Cases
def test_empty_json_string():
    codeflash_output = fix_unescaped_quotes_in_json('""')

def test_empty_json_object():
    codeflash_output = fix_unescaped_quotes_in_json('{}')

def test_empty_json_array():
    codeflash_output = fix_unescaped_quotes_in_json('[]')

def test_single_character_json_strings():
    codeflash_output = fix_unescaped_quotes_in_json('"a"')
    codeflash_output = fix_unescaped_quotes_in_json('"{"')
    codeflash_output = fix_unescaped_quotes_in_json('"}"')

# Complex Nested JSON Structures
def test_deeply_nested_json_object():
    codeflash_output = fix_unescaped_quotes_in_json('{"key1": {"key2": {"key3": {"key4": "value"}}}}')

def test_nested_json_arrays():
    codeflash_output = fix_unescaped_quotes_in_json('{"key": ["value1", ["value2", ["value3"]]]}')

# Invalid JSON Strings



def test_large_json_object():
    large_json = '{"' + '", "'.join(f'key{i}": "value{i}' for i in range(1000)) + '"}'
    fixed_large_json = '{"' + '", "'.join(f'key{i}": "value{i}' for i in range(1000)) + '"}'
    codeflash_output = fix_unescaped_quotes_in_json(large_json)

def test_large_json_array():
    large_json_array = '["' + '", "'.join(f'value{i}' for i in range(1000)) + '"]'
    fixed_large_json_array = '["' + '", "'.join(f'value{i}' for i in range(1000)) + '"]'
    codeflash_output = fix_unescaped_quotes_in_json(large_json_array)

def test_large_nested_json_structure():
    nested_json = '{"level1": {"level2": {"level3": {"level4": {"level5": {"level6": {"key": "deep value"}}}}}}}'
    codeflash_output = fix_unescaped_quotes_in_json(nested_json)
# codeflash_output is used to check that the output of the original code is the same as that of the optimized code.
```

</details>

To edit these changes `git checkout codeflash/optimize-fix_unescaped_quotes_in_json-m7ugbcmu` and push.

[Codeflash](https://codeflash.ai)

----

> [!IMPORTANT]
> Optimizes `fix_unescaped_quotes_in_json` in `utils.py` for 106% speedup by improving string construction and processing efficiency, with 52 regression tests added for validation.
> 
>   - **Performance Optimization**:
>     - `fix_unescaped_quotes_in_json` in `utils.py` optimized for 106% speedup.
>     - Avoids list sorting and index shifting by constructing the final string during traversal.
>     - Uses a list to accumulate characters, reducing string concatenation inefficiencies.
>     - Implements a single-pass approach for processing the input string.
>   - **Testing**:
>     - 52 regression tests added to verify correctness, covering various JSON string scenarios.
>     - Tests include valid JSON, escaped/unescaped quotes, nested structures, whitespace, edge cases, and special characters.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=Skyvern-AI%2Fskyvern&utm_source=github&utm_medium=referral)<sup> for 56230b57a7f98117b809d50c4e2195e5353ac033. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Review by wintonzheng — APPROVED** [review]

File: skyvern/forge/sdk/api/llm/utils.py [hunk 1]
```diff
@@ -146,23 +144,22 @@ def fix_unescaped_quotes_in_json(json_string: str) -> str:
...
         else:
             escape = False
+
+        # Append the current character to the result
+        result.append(char)
         i += 1
 
-    # Sort the indices in descending order to avoid index shifting then add the escape character to the string
-    if indices_to_add_escape_char:
+    if result != list(json_string):
```
[the remark this answers is not in the document]
Comment (wintonzheng) [reply skyvern/forge/sdk/api/llm/utils.py:159]:
```suggestion
    if len(result) != len(json_string):
```
